62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,753 @@
+            assert((candidatesSize % 2u)==0u);
+            for (auto i=0u; i< candidatesSize;)
+            {
+                const Segment& left = activeCandidates[i++];
+                const Segment& right = activeCandidates[i++];
+
+                CurveHatchBox curveBox;
+
+				// Due to precision, if the curve is right at the end, intersectOrtho may return nan
+				auto curveMinEnd = intersectOrtho(*left.originalBezier, newMajor, major);
+				auto curveMaxEnd = intersectOrtho(*right.originalBezier, newMajor, major);
+
+                auto splitCurveMin = splitCurveRange(*left.originalBezier, left.t_start, isnan(curveMinEnd) ? 1.0 : curveMinEnd);
+				auto splitCurveMax = splitCurveRange(*right.originalBezier, right.t_start, isnan(curveMaxEnd) ? 1.0 : curveMaxEnd);
+				assert(splitCurveMin.evaluate(0.0)[major] <= splitCurveMin.evaluate(1.0)[major]);
+				assert(splitCurveMax.evaluate(0.0)[major] <= splitCurveMax.evaluate(1.0)[major]);
+
+                auto curveMinAabb = getBezierBoundingBoxMinor(splitCurveMin);
+				auto curveMaxAabb = getBezierBoundingBoxMinor(splitCurveMax);
+                curveBox.aabbMin = float64_t2(std::min(curveMinAabb.first.x, curveMaxAabb.first.x), lastMajor);
+                curveBox.aabbMax = float64_t2(std::max(curveMinAabb.second.x, curveMaxAabb.second.x), newMajor);
+
+#ifdef DEBUG_HATCH_VISUALLY
+				if (isCurrentDebugStep)
+				{
+					drawDebugBezier(splitCurveMin, float64_t4(1.0, 0.0, 0.0, 1.0));
+					drawDebugBezier(splitCurveMax, float64_t4(0.0, 1.0, 0.0, 1.0));
+				}
+#endif
+
+                // Transform curves into AABB UV space and turn them into quadratic coefficients
+                // so we wont need to convert here
+                auto transformCurves = [](Hatch::QuadraticBezier bezier, float64_t2 aabbMin, float64_t2 aabbMax, float64_t2* output) {
+                    auto rcpAabbExtents = float64_t2(1.0,1.0) / (aabbMax - aabbMin);
+					// TODO: when nbl::hlsl::shapes::Quadratic<double> works, use it
+					auto transformedBezier = QuadraticBezier::construct(
+						(bezier.P0 - aabbMin) * rcpAabbExtents,
+						(bezier.P1 - aabbMin) * rcpAabbExtents,
+						(bezier.P2 - aabbMin) * rcpAabbExtents
+					);
+					auto quadratic = QuadraticEquation::constructFromBezier(transformedBezier);
+                    output[0] = quadratic.A;
+					output[1] = quadratic.B;
+					output[2] = quadratic.C;
+					if (isLineSegment(transformedBezier)) 
+						output[0].y = 0.0;
+					// B == 0.0 would mean this is a constant line in major direction, which
+					// should've been ruled out at this point (isStraightLineCosntantMajor gets skipped)
+					assert(quadratic.A.y > 0.0 || quadratic.B.y != 0.0);
+                };
+                transformCurves(splitCurveMin, curveBox.aabbMin, curveBox.aabbMax, &curveBox.curveMin[0]);
+                transformCurves(splitCurveMax, curveBox.aabbMin, curveBox.aabbMax, &curveBox.curveMax[0]);
+
+                hatchBoxes.push_back(curveBox);
```

Comment:
or you can do the packing somewhere around here and Hatch::CurveBox stores uint too
It'll be better cause no runtime coonversion will be needed and less memory to store for each box (cause we change 6 doubles to 6 uint32s)
but a bit harder to debug. make sure to unpack when working with your debugging values